nototools/gsub_diff.py

```diff
@@ -44,41 +44,144 @@ def __init__(self, file_a, file_b, output_lines=20):
-
-        rules_a = self._get_gsub_rules(self.text_a, self.file_a)
-        rules_b = self._get_gsub_rules(self.text_b, self.file_b)
-
-        diffs = []
-        report = ['']  # first line replaced by difference count
-        for rule in rules_a:
-            if rule not in rules_b:
-                diffs.append(('-',) + rule)
-        for rule in rules_b:
-            if rule not in rules_a:
-                diffs.append(('+',) + rule)
-        # ('+', 'smcp', 'Q', 'Q.sc')
+        new = [self._format_rule(r, '+') for r in self.find_new_rules()]
+        missing = [self._format_rule(r, '-') for r in self.find_missing_rules()]
+        diffs = missing + new
+        # ('+', 'smcp', 'Q', 'by', Q.sc')
         # Sort order:
         # 1. Feature tag
         # 2. Glyph name before substitution
         # 3. Glyph name after substitution
-        diffs.sort(key=lambda t:(t[1], t[2], t[3]))
+        diffs.sort(key=lambda t:(t[1], t[2], t[4]))
         report = ['%d differences in GSUB rules' % len(diffs)]
         report.extend(' '.join(diff) for diff in diffs)
         return '\n'.join(report[:self.output_lines + 1])
 
-    def _get_gsub_rules(self, text, filename):
-        """Get substitution rules in this ttxn output."""
+    def find_new_rules(self):
+        rules_a = self._get_gsub_rules(self.text_a, self.file_a)
+        rules_b = self._get_gsub_rules(self.text_b, self.file_b)
+        return [r for r in rules_b if r not in rules_a]
+
+    def find_missing_rules(self):
+        rules_a = self._get_gsub_rules(self.text_a, self.file_a)
+        rules_b = self._get_gsub_rules(self.text_b, self.file_b)
+        return [r for r in rules_a if r not in rules_b]
+
+    def _get_gsub_rules(self, text, file):
+        """
+        Parse the ttxn GSUB table in the following manner:
+
+        1. Get features
+        2. Get feature content
+        3. Extract lookup rules from feature content
+
+        Following substitutions are currently implemented:
+        - Type 1: Single substitutions
+        - Type 2: Multiple substitutions
+        - Type 3: Alternate substitutions
+        - Type 4: Ligature substitutionss
 
+        TODO: LookupTypes 5, 6, 8 still need implementing
+        """
+        rules = []
+        features = self._get_gsub_features(text)
+        for feature in features:
+            content = self._get_feature_content(text, feature)
+            lookups_rules = self._get_lookups_rules(text, content[0], feature)
+            rules += lookups_rules
+        return rules
+
+    def _get_gsub_features(self, text):
+        features = set()
         feature_name_rx = r'feature (\w+) {'
-        contents_rx = r'feature %s {(.*?)} %s;'
-        rule_rx = r'sub ([\w.]+) by ([\w.]+);'
 
-        rules = set()
         for name in re.findall(feature_name_rx, text):
-            contents = re.findall(contents_rx % (name, name), text, re.S)
-            assert len(contents) == 1, 'Multiple %s features in %s' % (
```

Comment:
the previous code would assert that only one feature block with a given name exists (which might not be true, at least for handwritten, non-ttxn-generated, feature files, since the spec allows multiple feature blocks with same name).
In your `_get_feature_content` below you have removed the assert and simply returning the result of `re.findall`, which may contain more than one match. But you are only retaining the first one, since you do `contents[0]` later on.
Was this intentional?
If we only support reading lookups from the first feature block that matches, then we need to assert that there's only one. Probably, it would be better to collect lookup rules for all the feature blocks with that name.